File: emarsys-sdk/src/main/java/com/emarsys/service/EmarsysMessagingService.java
package com.emarsys.service;

import com.emarsys.Emarsys;
import com.emarsys.core.di.DependencyContainer;
import com.emarsys.core.di.DependencyInjection;
import com.emarsys.mobileengage.di.MobileEngageDependencyContainer;
import com.google.firebase.messaging.FirebaseMessagingService;
import com.google.firebase.messaging.RemoteMessage;

public class EmarsysMessagingService extends FirebaseMessagingService {

    @Override
    public void onNewToken(String token) {
        super.onNewToken(token);
        if (DependencyInjection.getContainer().getDeviceInfo().isAutomaticPushSendingEnabled()) {
            Emarsys.Push.setPushToken(token);
        }
    }

    @Override
    public void onMessageReceived(final RemoteMessage remoteMessage) {
        super.onMessageReceived(remoteMessage);
        DependencyContainer container = DependencyInjection.<MobileEngageDependencyContainer>getContainer();
        container.getRunnerProxy().logException(new Runnable() {
            @Override
            public void run() {
                EmarsysMessagingServiceUtils.handleMessage(EmarsysMessagingService.this, remoteMessage);
            }
        });
    }
}

File: emarsys-sdk/src/main/java/com/emarsys/service/EmarsysMessagingService.kt
package com.emarsys.service

import com.emarsys.Emarsys
import com.emarsys.core.device.DeviceInfo
import com.emarsys.core.di.getDependency
import com.google.firebase.messaging.FirebaseMessagingService
import com.google.firebase.messaging.RemoteMessage

class EmarsysMessagingService : FirebaseMessagingService() {
    override fun onNewToken(token: String) {
        super.onNewToken(token)
        if (getDependency<DeviceInfo>().isAutomaticPushSendingEnabled) {
            Emarsys.push.setPushToken(token)
        }
    }

    override fun onMessageReceived(remoteMessage: RemoteMessage) {
        super.onMessageReceived(remoteMessage)
        EmarsysMessagingServiceUtils.handleMessage(this@EmarsysMessagingService, remoteMessage)
    }
}